src/dune_rules/findlib/findlib.ml

```diff
@@ -97,44 +104,164 @@ module Vars = struct
+        meta.vars |> String.Map.map ~f:Rules.of_meta_rules
+      in
+      let+ vars =
+        let* vars =
+          Fs_memo.file_exists config_file >>= function
+          | true -> load config_file
+          | false -> Memo.return Vars.empty
+        in
+        let config_dir =
+          Path.Outside_build_dir.extend_basename config_file ~suffix:".d"
+        in
+        Fs_memo.is_directory config_dir >>= function
+        | Ok true -> (
+          Fs_memo.dir_contents config_dir >>= function
+          | Ok dir_contents ->
+            let+ all_vars =
+              Memo.parallel_map (Fs_cache.Dir_contents.to_list dir_contents)
+                ~f:(fun (p, _kind) ->
+                  let p = Path.Outside_build_dir.relative config_dir p in
+                  load p)
+            in
+            List.fold_left all_vars ~init:vars ~f:(fun acc vars ->
+                Vars.union acc vars ~f:(fun _ x y -> Some (Rules.union x y)))
+          | Error _ -> Memo.return vars)
+        | Ok false | Error _ -> Memo.return vars
+      in
+      { vars; preds = Ps.empty }
+
+    let get { vars; preds } var = Vars.get vars var preds
+
+    let toolchain t ~toolchain =
+      { t with preds = Ps.singleton (P.make toolchain) }
+  end
+
   type t =
-    { vars : Vars.t
-    ; preds : Ps.t
+    { config : File.t
+    ; ocamlpath : Path.t list
+    ; which : string -> Path.t option Memo.t
+    ; toolchain : string option
     }
 
-  let to_dyn { vars; preds } =
+  let to_dyn { config; ocamlpath; toolchain; which = _ } =
     let open Dyn in
     record
-      [ ("vars", String.Map.to_dyn Rules.to_dyn vars)
-      ; ("preds", Ps.to_dyn preds)
+      [ ("config", File.to_dyn config)
+      ; ("ocamlpath", Dyn.list Path.to_dyn ocamlpath)
+      ; ("toolchain", option string toolchain)
       ]
 
-  let load path ~toolchain ~context =
-    let path = Path.Outside_build_dir.extend_basename path ~suffix:".d" in
-    let conf_file =
-      Path.Outside_build_dir.relative path (toolchain ^ ".conf")
-    in
-    let* conf_file_exists = Fs_memo.file_exists conf_file in
-    if not conf_file_exists then
-      User_error.raise
-        [ Pp.textf "ocamlfind toolchain %s isn't defined in %s (context: %s)"
-            toolchain
-            (Path.Outside_build_dir.to_string_maybe_quoted path)
-            context
-        ];
-    let+ meta = Meta.load ~name:None conf_file in
-    { vars = String.Map.map meta.vars ~f:Rules.of_meta_rules
-    ; preds = Ps.of_list [ P.make toolchain ]
-    }
+  let ocamlpath_sep =
+    if Sys.cygwin then (* because that's what ocamlfind expects *)
+      ';'
+    else Bin.path_sep
 
-  let get { vars; preds } var = Vars.get vars var preds
+  let path_var = Bin.parse_path ~sep:ocamlpath_sep
 
-  let env t =
-    let preds = Ps.add t.preds (P.make "env") in
-    String.Map.filter_map ~f:(Rules.interpret ~preds) t.vars
+  let ocamlpath env =
+    match Env.get env "OCAMLPATH" with
+    | None -> []
```

Comment:
Is it ok that we treat the empty OCAMLPATH variable the same as not setting it at all?

Reply:
It might be a good idea to revert this to returning an option type. I'll try that.